File: 516. Longest Palindromic Subsequence/dp.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        N = len(s)
        prev, curr = [0] * N, [0] * N;
        res = 0;
        
        for i in reversed(range(N)):
            for j in range(i, N):
                if i == j:
                    curr[j] = 1
                else:
                    if s[i] == s[j]:
                        if i + 1 == j:
                            curr[j] = 2;
                        else:
                            curr[j] = 2 + prev[j-1];
                    else:
                        if i + 1 == j :
                            curr[j] = 1;
                        else:
                            curr[j] = max(curr[j-1], prev[j])
                            
                res = max(res, curr[j])
            
            (prev, curr) = (curr, prev);
        
        return res;
        
                        

File: 516. Longest Palindromic Subsequence/dp.js
/**
 * @param {string} s
 * @return {number}
 */
 var longestPalindromeSubseq = function(s) {
    const n = s.length;
    
    const dp = [...Array(n)].map(x => [...Array(n)].fill(0))
    let max = 0
    
    for(let i = n-1; i >= 0; i--){
        for(let j = i; j < n; j++){
            if(i === j){
                dp[i][j] = 1;             
            }else{
                if(s[i] === s[j]){
                    if(i + 1 === j){
                        dp[i][j] = 2
                    }else{
                        dp[i][j] = 2 + dp[i+1][j-1]
                    }
                }else{
                    if(i + 1 === j) dp[i][j] = 1;
                    else dp[i][j] = Math.max(dp[i+1][j], dp[i][j-1])
                }
            }
            max = Math.max(max, dp[i][j])
        }
    }
    
    return max;
};